### PR TITLE
jbide-21608: disabling "Show In -> Web Browser via Live Reload" on EJB module

### PR DESCRIPTION
**Pull request by bdshadow**

### CODE REVIEW

File: plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java [hunk 1]
```diff
@@ -74,6 +73,7 @@
 	public static final String JBOSSAS_SERVER_PORT = "org.jboss.ide.as.serverport";
 	public static final String TOMCAT_60_SERVER_TYPE = "org.eclipse.jst.server.tomcat.60";
 	public static final String TOMCAT_70_SERVER_TYPE = "org.eclipse.jst.server.tomcat.70";
+	public static final String EJB_MODULE_TYPE = "jst.ejb";
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: Is this constant available somewhere in JST for us to reuse? If yes, we should reuse the original constant instead.

Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: Didn't found it in jst, but in org.jboss.ide.eclipse.as.core.util.IWTPConstants. Replaced

**Comment by xcoulon** [comment]
@bdshadow this looks good to me at first sight. Could you add some tests to verify that the command is enabled for web modules and disabled on non web modules ?

**Comment by bdshadow** [comment]
@xcoulon need to clarify here a bit. Web modules are not ejb modules. They have different 'IModuleType's: jst.ejb and jst.web. In JIRA task description it is written to disable the menu items for ejb modules, as i've done. So in my opinion the same thing, done by @robstryker  [here](https://github.com/jbosstools/jbosstools-server/blob/5330c96aa7ad25fa1a5fa137c8635272245a9ef9/as/plugins/org.jboss.ide.eclipse.as.ui/jbossui/org/jboss/ide/eclipse/as/ui/views/server/extensions/ShowInWelcomePageActionProvider.java#L250) and in [this PR](https://github.com/jbosstools/jbosstools-server/pull/386) is not totally correct. I debugged this server code and the modules were filtered not because they didn't have web-address to open but because they are  of different 'IModuleType'.

**Comment by robstryker** [comment]
The PR you've linked @bdshadow  is for the welcome page, ie, when right-clicking on the server (not a child module). So that was for a different issue.

File: plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java [hunk 2]
```diff
@@ -571,5 +568,39 @@ private static String getContextRoot(final IServer server, final IModule module)
...
+			return null;
+		}
+		final IURLProvider urlProvider = getURLProvider(server);
+		if (urlProvider != null) {
+			return urlProvider.getModuleRootURL(module);
+		}
+		return null;
+	}
+	
+	private static boolean isEjbModule(IModule module) {
```
Comment (xcoulon) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: After some thoughs, shouldn't we check that the module is a Web module instead of excluding EJB modules ?

File: plugins/org.jboss.tools.livereload.ui/plugin.xml [hunk 3]
```diff
@@ -82,9 +82,9 @@
                  <iterate
                        ifEmpty="false"
                        operator="or">
-                    	<adapt
-							type="org.eclipse.wst.server.ui.IServerModule">
-						</adapt>
+                    	<instanceof
+							value="org.eclipse.wst.server.ui.IServerModule">
+						</instanceof>
```
Comment (xcoulon) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: why this change ? what does it fix ?

File: plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java [hunk 4]
```diff
@@ -34,7 +35,10 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
 			return false;
 		}
 		final IServer appServer = appModule.getServer();
-		return appServer != null && WSTUtils.isServerStarted(appServer);
+		IModule[] modules = appModule.getModule();
+		return appServer != null
+				&& WSTUtils.isServerStarted(appServer)
+				&& WSTUtils.getModuleURL(appServer, modules[modules.length - 1]) != null;
 	}
```
Comment (xcoulon) [inline plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java]: since you're adding more conditions to enable the command (which seems legit), could you add a set of tests to verify that it works as expected in all situations ? Eg: when the project is a .war or an .ear with web modules and ejb modules, started or stopped.

File: plugins/org.jboss.tools.livereload.ui/plugin.xml (hunk 3 above)
Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: @xcoulon it makes commands `org.jboss.tools.livereload.openInExternalDeviceWebBrowserViaQRCode` and `org.jboss.tools.livereload.openInWebBrowserViaLiveReloadProxy` enabled identically (see line 109). By now, without this changes, "Open in web browser via LiveReload" was enabled while "Open in External Device" was disabled, because `adapt` is a bit more than `instanceof`.

File: plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java (hunk 2 above)
Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: @xcoulon maybe. I think we need some more opinions. Anyway we need to make it everywhere the same way (see links of [the comment](https://github.com/jbosstools/jbosstools-livereload/pull/129#issuecomment-227119501) above)

File: plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java [hunk 5]
```diff
@@ -34,7 +35,11 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
 			return false;
 		}
 		final IServer appServer = appModule.getServer();
-		return appServer != null && WSTUtils.isServerStarted(appServer);
+		IModule[] modules = appModule.getModule();
+		return appServer != null
+				&& appServer.getModuleState(appModule.getModule()) == IServer.STATE_STARTED
```
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java]: When a web module is deployed within an EAR (try wildfly-javaee7-webapp-ear-archetype), on WF10, it's state is IServer.STATE_UNKNOWN, so the tester returns false here and it makes it impossible to open via livereload

File: plugins/org.jboss.tools.livereload.ui/plugin.xml (hunk 3 above)
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: this change was applied in https://github.com/jbosstools/jbosstools-livereload/commit/67d2360fe1ed4478c8fbe200a6c00b9780520022. Please rebase/squash.

Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: @fbricon github says: "This branch has no conflicts with the base branch"

File: plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java (hunk 5 above)
Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java]: @fbricon, see [this JIRA](https://issues.jboss.org/browse/JBIDE-21936). It was said to disable for stopped modules. So i've made a check that module is started. Do we really need to make it enabled for unknown state?

Comment (fbricon) [inline plugins/org.jboss.tools.livereload.ui/src/org/jboss/tools/livereload/ui/internal/command/OpenInWebBrowserViaLiveReloadProxyEnablementTester.java]: In doubt (unknown state) then yes

File: plugins/org.jboss.tools.livereload.ui/plugin.xml (hunk 3 above)
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: I know, but it will be less confusing when accessing the PR on GH.

**Comment by fbricon** [comment]
Tycho tests fail with

```
Failed tests:
  WSTUtilsTestCase.shouldFindLiveReloadServerForPreviewServer:211
Expecting actual not to be null
  WSTUtilsTestCase.shouldGetModuleContext:261
Expecting:
 <"http://hostname:8080/">
to be equal to:
 <"http://hostname:8080/module/">
but was not.
```

File: plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java [hunk 6]
```diff
@@ -571,5 +568,39 @@ private static String getContextRoot(final IServer server, final IModule module)
...
+	 * Gets the {@link URL} of the deployed {@code module}  deployed on the given {@code server}.
+	 * @param server the {@link IServer} on which the module is deployed
+	 * @param module the {@link IModule} whose context root is to be found
+	 * @return the URL of the {@code module} on the {@code server} or
+	 * 		   {@code null} if it is not a web module
+	 * 					    or if this server does not have a valid configuration
+	 */
+	public static URL getModuleURL(IServer server, IModule module) {
+		ServerExtendedProperties props = getServerExtendedProperties(server);
+		if (props == null || isEjbModule(module)) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: WSTUtilsTestCase .shouldGetModuleContext fails because the tested project is an ejb, so the returned URL is null at this point

Comment (bdshadow) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: @fbricon this was asked in the task https://issues.jboss.org/browse/JBIDE-21608. Ans @robstryker did it  [the same way](https://github.com/jbosstools/jbosstools-server/blob/5330c96aa7ad25fa1a5fa137c8635272245a9ef9/as/plugins/org.jboss.ide.eclipse.as.ui/jbossui/org/jboss/ide/eclipse/as/ui/views/server/extensions/ShowInWelcomePageActionProvider.java#L239) for the relative JIRA (https://issues.jboss.org/browse/JBIDE-21138). He saw this PR, said it was ok

Comment (olliebeef) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: Well I'm seeing conflicting reqs here.  One is saying it's supposed to be disabled for ejb. But a unit test is testing it against an ejb.  So I'd argue the test is wrong?

File: plugins/org.jboss.tools.livereload.ui/plugin.xml (hunk 3 above)
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.ui/plugin.xml]: So actually this changes breaks openshift support. The menu is enabled at the openshift server level, which is not an instance of IServerModule, but adapts to IServerModule. It was broken in openInExternalDeviceWebBrowserViaQRCode, which was fixed in 67d2360

**Comment by fbricon** [comment]
@bdshadow using stating imports on Mockito.* will reduce some clutter

File: plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java [hunk 7]
```diff
@@ -571,5 +568,39 @@ private static String getContextRoot(final IServer server, final IModule module)
...
+		final IURLProvider urlProvider = getURLProvider(server);
+		if (urlProvider != null) {
+			return urlProvider.getModuleRootURL(module);
+		}
+		return null;
+	}
+	
+	private static boolean isEjbModule(IModule module) {
+		IModuleType type = module.getModuleType();
+		return IWTPConstants.FACET_EJB.equals(type.getId()); //$NON-NLS-1$
```
Comment (fbricon) [inline plugins/org.jboss.tools.livereload.core/src/org/jboss/tools/livereload/core/internal/util/WSTUtils.java]: using an openshift server:
```
java.lang.NullPointerException
  at org.jboss.tools.livereload.core.internal.util.WSTUtils.isEjbModule(WSTUtils.java:603)
  at org.jboss.tools.livereload.core.internal.util.WSTUtils.getModuleURL(WSTUtils.java:591)
```
type is null for openshift servers
